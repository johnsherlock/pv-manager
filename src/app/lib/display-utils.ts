export const isTouchScreen = () => 'ontouchstart' in window || navigator.maxTouchPoints > 0;