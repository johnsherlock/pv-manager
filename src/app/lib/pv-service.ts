import axios from 'axios';
import { EddiData } from '../../shared/eddi-data';
import { mapEddiDataToMinutePVData } from '../../shared/energy-utils';
import { MinutePVData, Totals } from '../../shared/pv-data';

export interface DailyEnergyUsageProps {
  data: MinutePVData[];
}

export const getPVDataForDate = async (formattedTargetDate: string): Promise<MinutePVData[]> => {
  console.log(`Retrieving per minute data for ${formattedTargetDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/minute-data?date=${formattedTargetDate}`;
  try {
    const response = await axios.get(url);
    const data = await Promise.resolve(response.data as EddiData[]);
    return data.map(mapEddiDataToMinutePVData);
  } catch (error) {
    console.log(`Error retrieving data for ${formattedTargetDate}`, error);
    document.body.style.cursor = 'auto';
    return [];
  };
};

export const getPVTotalsForRange = async (startDate: string, endDate: string): Promise<Totals[]> => {
  console.log(`Retrieving PV summary data for range ${startDate} - ${endDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/aggregate-historical-data?serialNumber=21494842&startDate=${startDate}&endDate=${endDate}`;
  try {
    const response = await axios.get(url);
    return await Promise.resolve(response.data as Totals[]);
  } catch (error) {
    console.log('Error retrieving totals for range', error);
    document.body.style.cursor = 'auto';
    return [];
  };
};