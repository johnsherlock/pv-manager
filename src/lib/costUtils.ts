import * as numUtils from './numUtils';

// export discounted rates
// export const dayRate = 0.4330;
// export const peakRate = 0.5289;
// export const nightRate = 0.3182;

// export 100% standard rates from bill
export const dayRate = 0.4673;
export const peakRate = 0.5709;
export const nightRate = 0.3434;

// export calculated 100% standard rates
// export const dayRate = 0.5094;
// export const peakRate = 0.6222;
// export const nightRate = 0.3742;

export const exportRate = 0.1850;

export const discount = 1 - 0.15;
export const standingCharge = 0.7066;
export const hourlyStandingCharge = standingCharge / 24;
export const vatRate = 1.09;

export const formatToEuro = (amount: number) => (amount ? `â‚¬${amount.toFixed(2)}` : '');

export const calculateHourlyNetCostAtStandardRates = (hour = 0, dow: string, joules: number) => {
  if (joules) {
    let multiplier = dayRate;

    if ((hour >= 0 && hour <= 8) || hour === 23) {
      multiplier = nightRate;
    } else if (hour >= 17 && hour < 19) {
      multiplier = peakRate;
    }

    if (dow === 'Sat' && hour >= 9 && hour < 17) {
      multiplier = 0;
    }

    const kWh = numUtils.convertJoulesToKwh(joules);
    const cost = kWh * multiplier;

    return numUtils.formatDecimal(cost);
  }
  return 0;
};

export const calculateDiscountedHourlyGrossCost = (hour: number = 0, dow: string, joules: number) => {
  const netCostAtStandardRates = calculateHourlyNetCostAtStandardRates(hour, dow, joules);
  const grossCost = (netCostAtStandardRates * discount) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateHourlyGrossCostIncStdChgAndDiscount = (hour: number = 0, dow: string, joules: number) => {
  const netCostAtStandardRates = calculateHourlyNetCostAtStandardRates(hour, dow, joules);
  const grossCost = ((netCostAtStandardRates * discount) + hourlyStandingCharge) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateSaturdaySaving = (hour = 0, dow: string, joules: number) => {
  if (joules && dow === 'Sat' && hour >= 9 && hour <= 17) {
    const kWh = numUtils.convertJoulesToKwh(joules);
    const netSavingAtStandardRates = kWh * dayRate;
    const grossSavingAtDiscountedRates = (netSavingAtStandardRates * discount) * vatRate;
    return numUtils.formatDecimal(grossSavingAtDiscountedRates);
  }
  return 0;
};

export const calculateDiscountedGrossCostExcludingStdChg = (netCostAtStandardRate: number) => {
  const grossCost = (netCostAtStandardRate * discount) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateGrossCostIncStandingCharges = (netCost: number) => {
  const grossCost = ((netCost * discount) + hourlyStandingCharge) * vatRate;
  return numUtils.formatDecimal(grossCost);
};

export const calculateExportValue = (joules: number) => {
  if (joules) {
    const kWh = numUtils.convertJoulesToKwh(joules);
    return numUtils.formatDecimal(kWh * exportRate);
  }
  return 0;
};
