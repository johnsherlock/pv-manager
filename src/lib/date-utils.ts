import moment from 'moment-timezone';
import { BasePVData } from '../model/pv-data';

export const formatDate = (date: moment.Moment): string => moment(date).startOf('day').format('YYYY-MM-DD');

export const dawnOfTime = moment(new Date('2023-01-20')).startOf('day');

export const toMoment = (pvData: BasePVData, locale?: string): moment.Moment => {
  return moment({
    year: pvData.year,
    month: pvData.month - 1,
    date: pvData.dayOfMonth,
    hour: pvData.hour,
    minute: pvData.minute,
  });
};

export const getFormattedTime = (pvData: BasePVData, locale?: string): string => {
  const time = toMoment(pvData, locale);
  return time.format('HH:mm');
};