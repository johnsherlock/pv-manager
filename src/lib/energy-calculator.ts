import * as numUtils from './num-utils';
import { MinutePVData, HalfHourlyPVData, calculateGreenEnergyPercentage } from './pv-service';

export interface EnergyCalculatorProps {
  readonly dayRate: number;
  readonly peakRate: number;
  readonly nightRate: number;
  readonly exportRate: number;
  readonly discountPercentage: number;
  readonly annualStandingCharge: number;
  readonly vatRate?: number;
}

export interface Totals {
  impTotal: number;
  genTotal: number;
  expTotal: number;
  conpTotal: number;
  dayImpTotal: number;
  peakImpTotal: number;
  nightImpTotal: number;
  freeImpTotal: number;
  immersionTotal: number;
  immersionRunTime: number;
  grossSavingTotal: number;
}

const initialTotals = (): Totals => {
  return {
    impTotal: 0,
    genTotal: 0,
    expTotal: 0,
    conpTotal: 0,
    dayImpTotal: 0,
    peakImpTotal: 0,
    nightImpTotal: 0,
    freeImpTotal: 0,
    immersionRunTime: 0,
    immersionTotal: 0,
    grossSavingTotal: 0,
  };
};

export class EnergyCalculator {

  readonly dayRate: number;
  readonly peakRate: number;
  readonly nightRate: number;
  readonly exportRate: number;
  readonly discountPercentage: number;
  readonly annualStandingCharge: number;
  readonly dailyStandingCharge: number;
  readonly hourlyStandingCharge: number;
  readonly halfHourlyStandingCharge: number;
  readonly perMinuteStandingCharge: number;
  readonly vatRate: number;

  constructor(props: EnergyCalculatorProps) {
    this.dayRate = props.dayRate;
    this.peakRate = props.peakRate;
    this.nightRate = props.nightRate;
    this.exportRate = props.exportRate;
    this.discountPercentage = 1 - props.discountPercentage;
    this.annualStandingCharge = props.annualStandingCharge;
    this.dailyStandingCharge = props.annualStandingCharge / 365;
    this.hourlyStandingCharge = this.dailyStandingCharge / 24;
    this.halfHourlyStandingCharge = this.hourlyStandingCharge / 2;
    this.perMinuteStandingCharge = this.halfHourlyStandingCharge / 30;
    this.vatRate = props.vatRate ?? 1.09;
  }

  public calculateNetCostAtStandardRates = (hour = 0, dow: string, kWh: number = 0, round: boolean = true) => {
    let multiplier = this.dayRate;

    if ((hour >= 0 && hour <= 8) || hour === 23) {
      multiplier = this.nightRate;
    } else if (hour >= 17 && hour < 19) {
      multiplier = this.peakRate;
    }

    if (dow === 'Sat' && hour >= 9 && hour < 17) {
      multiplier = 0;
    }

    const cost = kWh * multiplier;

    return round ? numUtils.formatDecimal(cost) : cost;
  };

  public calculateDiscountedCostIncludingVat = (hour: number = 0, dow: string, kWh: number = 0, round = true) => {
    const netCostAtStandardRates = this.calculateNetCostAtStandardRates(hour, dow, kWh, round);
    const grossCost = (netCostAtStandardRates * this.discountPercentage) * this.vatRate;
    return round ? numUtils.formatDecimal(grossCost) : grossCost;
  };

  public calculateSaving = (hour: number = 0, dow: string, importedkWh: number = 0, consumedkWh: number = 0, round = true) => {
    const greenkWh = consumedkWh - importedkWh;
    return greenkWh > 0 ? this.calculateDiscountedCostIncludingVat(hour, dow, greenkWh, round) : 0;
  };

  public calculateGrossCostPerHourIncStdChgAndDiscount = (hour: number = 0, dow: string, kWh: number = 0, round = true) => {
    return this.calculateGrossCostIncStdChgAndDiscount(hour, dow, kWh, 'hr', round);
  };

  public calculateGrossCostPerHalfHourIncStdChgAndDiscount = (hour: number = 0, dow: string, kWh: number = 0, round = true) => {
    return this.calculateGrossCostIncStdChgAndDiscount(hour, dow, kWh, 'hhr', round);
  };

  public calculateGrossCostPerMinuteIncStdChgAndDiscount = (hour: number = 0, dow: string, kWh: number = 0, round = false) => {
    return this.calculateGrossCostIncStdChgAndDiscount(hour, dow, kWh, 'm', round);
  };

  public calculateGrossCostIncStdChgAndDiscount = (hour: number = 0, dow: string, kWh: number = 0, timeUnit: 'hr' | 'hhr' | 'm', round = true) => {
    const netCostAtStandardRates = this.calculateNetCostAtStandardRates(hour, dow, kWh, round);
    const standingCharge = timeUnit === 'hr' ? this.hourlyStandingCharge : timeUnit === 'hhr' ? this.halfHourlyStandingCharge : this.perMinuteStandingCharge;
    const grossCost = ((netCostAtStandardRates * this.discountPercentage) + standingCharge) * this.vatRate;
    return round ? numUtils.formatDecimal(grossCost) : grossCost;
  };

  public calculateSaturdaySaving = (hour = 0, dow: string, kWh: number = 0) => {
    if (kWh && dow === 'Sat' && hour >= 9 && hour <= 17) {
      const netSavingAtStandardRates = kWh * this.dayRate;
      const grossSavingAtDiscountedRates = (netSavingAtStandardRates * this.discountPercentage) * this.vatRate;
      return numUtils.formatDecimal(grossSavingAtDiscountedRates);
    }
    return 0;
  };

  public calculateDiscountedGrossCostExcludingStdChg = (netCostAtStandardRate: number = 0) => {
    const grossCost = (netCostAtStandardRate * this.discountPercentage) * this.vatRate;
    return numUtils.formatDecimal(grossCost);
  };

  public calculateGrossCostIncStandingCharges = (netCost: number = 0) => {
    const grossCost = ((netCost * this.discountPercentage) + this.hourlyStandingCharge) * this.vatRate;
    return numUtils.formatDecimal(grossCost);
  };

  public calculateExportValue = (exportKwH: number = 0) => {
    return numUtils.formatDecimal(exportKwH * this.exportRate);
  };

  public calculateDailyGrossImportTotal = (totals: Totals = initialTotals()) => {
    const discountedDayImportNetCost =
      (numUtils.convertJoulesToKwh(totals.dayImpTotal - totals.freeImpTotal) * this.dayRate) * this.discountPercentage;
    const discountedPeakImportNetCost = (numUtils.convertJoulesToKwh(totals.peakImpTotal) * this.peakRate) * this.discountPercentage;
    const discountedNightImportNetCost = (numUtils.convertJoulesToKwh(totals.nightImpTotal) * this.nightRate) * this.discountPercentage;
    const discountedNetImportTotal = discountedDayImportNetCost + discountedPeakImportNetCost + discountedNightImportNetCost;
    const grossImportTotal = (discountedNetImportTotal + this.dailyStandingCharge) * this.vatRate;
    return numUtils.formatToEuro(grossImportTotal);
  };

  // public calculateFreeImportGrossTotal = (totals: Totals = initialTotals()) => {
  //   const freeImportGrossTotal = (numUtils.convertJoulesToKwh(totals.freeImpTotal) * this.dayRate) * this.vatRate;
  //   return numUtils.formatToEuro(freeImportGrossTotal);
  // };

  public calculateDailyExportTotal = (totals: Totals = initialTotals()) => {
    const grossExportTotal = (numUtils.convertJoulesToKwh(totals.expTotal) * this.exportRate);
    return numUtils.formatToEuro(grossExportTotal);
  };

  public calculateDailyGreenEnergyCoverage = (totals: Totals = initialTotals()) => {
    return numUtils.formatDecimal(100 - ((totals.impTotal/totals.conpTotal) * 100));
  };

  public calculateTotalImportedKwH = (pvData: HalfHourlyPVData[] = []) => {
    return numUtils.formatDecimal(pvData.reduce((acc, item) => acc + item.importedKwH, 0));
  };

  public calculateTotalGeneratedKwH = (pvData: HalfHourlyPVData[] = []) => {
    return numUtils.formatDecimal(pvData.reduce((acc, item) => acc + item.generatedKwH, 0));
  };

  public calculateTotalConsumedKwH = (pvData: HalfHourlyPVData[] = []) => {
    return numUtils.formatDecimal(pvData.reduce((acc, item) => acc + item.consumedKwH, 0));
  };

  public calculateTotalExportedKwH = (pvData: HalfHourlyPVData[] = []) => {
    return numUtils.formatDecimal(pvData.reduce((acc, item) => acc + item.exportedKwH, 0));
  };

  public calculateTotalImmersionDivertedKwH = (pvData: HalfHourlyPVData[] = []) => {
    return numUtils.formatDecimal(pvData.reduce((acc, item) => acc + item.immersionDivertedKwH, 0));
  };

  public calculateTotalImmersionBoostedKwH = (pvData: HalfHourlyPVData[] = []) => {
    return numUtils.formatDecimal(pvData.reduce((acc, item) => acc + (item.immersionBoostedKwH ?? 0), 0));
  };

  public calculateTotalImmersionDivertedMins = (pvData: HalfHourlyPVData[] = []) => {
    return pvData.reduce((acc, item) => acc + (item.immersionDivertedMins ?? 0), 0);
  };

  public calculateTotalImmersionBoostedMins = (pvData: HalfHourlyPVData[] = []) => {
    return pvData.reduce((acc, item) => acc + (item.immersionBoostedMins ?? 0), 0);
  };

  public calculateTotalGrossImportCost = (pvData: HalfHourlyPVData[] = []) => {
    return pvData.reduce((acc, item) => acc + this.calculateGrossCostPerHalfHourIncStdChgAndDiscount(item.hour, item.dayOfWeek, item.importedKwH), 0);
  };

  public calculateTotalGrossSavings = (pvData: HalfHourlyPVData[] = []) => {
    return pvData.reduce((acc, item) => acc + this.calculateSaving(item.hour, item.dayOfWeek, item.importedKwH, item.consumedKwH), 0);
  };

  public calculateTotalExportValue = (pvData: HalfHourlyPVData[] = []) => {
    return pvData.reduce((acc, item) => acc + this.calculateExportValue(item.exportedKwH), 0);
  };

  public calculateFreeImportGrossTotal = (pvData: HalfHourlyPVData[] = []) => {
    const freeImportedKwh: HalfHourlyPVData[] = pvData.filter(item => item.dayOfWeek === 'Sat' && item.hour >= 9 && item.hour < 17);
    return freeImportedKwh.reduce((acc, item) => acc + this.calculateDiscountedCostIncludingVat(item.hour, item.dayOfWeek, item.importedKwH), 0);
  };

  public calculaterTotalGreenEnergyCoverage = (pvData: HalfHourlyPVData[] = []) => {
    const totalImportedKwH = this.calculateTotalImportedKwH(pvData);
    const totalConsumedKwH = this.calculateTotalConsumedKwH(pvData);
    return calculateGreenEnergyPercentage(totalImportedKwH, totalConsumedKwH);
  };

  public recalculateTotals = (perMinuteData: MinutePVData[]): Totals => {
    console.log('Recaculating totals');
    const totals: Totals = initialTotals();
    perMinuteData.forEach((item: MinutePVData) => {
      totals.impTotal += item.importedJoules ?? 0;
      totals.genTotal += item.generatedJoules ?? 0;
      totals.expTotal += item.exportedJoules ?? 0;
      totals.conpTotal += item.consumedJoules ?? 0;
      totals.immersionRunTime += (item.immersionBoostedJoules || item.immersionDivertedJoules) ? 1 : 0;
      totals.immersionTotal += item.immersionDivertedJoules ?? 0;
      totals.grossSavingTotal += this.calculateSaving(
        item.hour, item.dayOfWeek, item.importedJoules, item.consumedJoules, false);

      if (item.hour >= 17 && item.hour < 19) totals.peakImpTotal += item.importedJoules ?? 0;
      else if ((item.hour >= 0 && item.hour < 8) || item.hour === 23) totals.nightImpTotal += item.importedJoules ?? 0;
      else if (item.dayOfWeek === 'Sat' && item.hour >= 9 && item.hour < 17) totals.freeImpTotal += item.importedJoules ?? 0;
      else totals.dayImpTotal += item.importedJoules ?? 0;
    });
    return {
      ...totals,
      impTotal: totals.peakImpTotal + totals.nightImpTotal + totals.dayImpTotal,
    };
  };

}