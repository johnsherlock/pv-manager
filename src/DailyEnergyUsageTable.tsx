import * as costUtils from './lib/costUtils';
import { Totals } from './lib/costUtils';
import * as numUtils from './lib/numUtils';

export interface DailyEnergyUsageTableProps {
  data: any;
  totals: Totals;
}

const DailyEnergyUsageTable = ({ data, totals }: DailyEnergyUsageTableProps): JSX.Element => {

  return (
    <div className="table">
      <div className="table-header">
        <div className="table-row">
          <div className="table-cell">Hour</div>
          <div className="table-cell">Imported kWh</div>
          <div className="table-cell">Cost</div>
          <div className="table-cell">Generated kWh</div>
          <div className="table-cell">Saving</div>
          <div className="table-cell">Exported kWh</div>
          <div className="table-cell">Value</div>
        </div>
      </div>
      <div className="table-container">
        <div className="table-body">
          {data?.map((item: any, index: number) => (
            <div key={item.yr + item.mon + item.dom + item.hr} className={`table-row ${index % 2 === 0 ? 'table-primary' : ''}`}>
              <div className="table-cell">{item.hr ? item.hr.toString().padStart(2, '0') : '00'}</div>
              <div className="table-cell">{numUtils.formatDecimal(numUtils.convertJoulesToKwh(item.imp))}</div>
              <div className="table-cell">{costUtils.formatToEuro(costUtils.calculateHourlyGrossCostIncStdChgAndDiscount(item.hr, item.dow, item.imp)) ?? ''}</div>
              <div className="table-cell">{numUtils.formatDecimal(numUtils.convertJoulesToKwh(item.gep)) ?? ''}</div>
              <div className="table-cell">{costUtils.formatToEuro(costUtils.calculateDiscountedHourlyGrossCost(item.hr, item.dow, item.gep))}</div>
              <div className="table-cell">{numUtils.formatDecimal(numUtils.convertJoulesToKwh(item.exp))}</div>
              <div className="table-cell">{numUtils.formatDecimal(costUtils.calculateExportValue(item.exp))}</div>
            </div>
          ))}
        </div>
      </div>
      <div className="table-footer">
        <div className="table-row">
          <span className="table-cell">Total</span>
          <span className="table-cell">
            {numUtils.formatDecimal(
              numUtils.convertJoulesToKwh(totals.impTotal))}
            {' '}
            kWh
          </span>
          <span className="table-cell">
            {costUtils.formatToEuro(totals?.grossCostTotal)}
            &nbsp;
            {totals?.saturdayNetSavingTotal ? `(${costUtils.formatToEuro(costUtils.calculateDiscountedGrossCostExcludingStdChg(totals?.saturdayNetSavingTotal))})` : ''}
          </span>
          <span className="table-cell">
            {numUtils.formatDecimal(numUtils.convertJoulesToKwh(totals?.genTotal))}
            {' '}
            kWh
          </span>
          <span className="table-cell">{costUtils.formatToEuro(totals?.grossSavingTotal) || '€0.00'}</span>
          <span className="table-cell">
            {numUtils.formatDecimal(numUtils.convertJoulesToKwh(totals?.expTotal))}
            {' '}
            kWh
          </span>
          <span className="table-cell">{costUtils.formatToEuro(totals?.exportValueTotal) || '€0.00'}</span>
        </div>
      </div>
    </div>

  );
};

export default DailyEnergyUsageTable;