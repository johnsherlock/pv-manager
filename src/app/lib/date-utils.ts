import moment from 'moment';
import { CalendarScale, DateRange, FormattedDateRange } from './state-utils';
import { BasePVData } from '../../shared/pv-data';

export const formatDate = (date: moment.Moment): string => moment(date).startOf('day').format('YYYY-MM-DD');

export const dawnOfTime = moment(new Date('2023-01-20')).startOf('day');

export const toMoment = (pvData: BasePVData, locale?: string): moment.Moment => {
  return moment({
    year: pvData.year,
    month: pvData.month - 1,
    date: pvData.dayOfMonth,
    hour: pvData.hour,
    minute: pvData.minute,
  });
};

export const getFormattedTime = (pvData: BasePVData, locale?: string): string => {
  const time = toMoment(pvData, locale);
  return time.format('HH:mm');
};

export const getDateRange = (scale: CalendarScale, date: Date): DateRange => {

  switch (scale) {
    case 'week':
    case 'month':
    case 'year': {
      const m = moment(date);
      const start = moment(m.startOf(scale).toDate());
      const end = moment(m.endOf(scale).toDate());
      return { startDate: start, endDate: end };
    }
    default:
      return { startDate: moment(date), endDate: (moment(date)) };
  }
};

export const formatDateRange = (dateRange: DateRange): FormattedDateRange => {
  return {
    startDate: formatDate(dateRange.startDate),
    endDate: formatDate(dateRange.endDate),
  };
};