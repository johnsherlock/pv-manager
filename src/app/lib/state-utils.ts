import moment from 'moment';
import * as dateUtils from './date-utils';
import { MinutePVData, Totals } from '../../shared/pv-data';
import { Scale, View } from '../energy-usage-line-graph';

export type CalendarScale = 'day' | 'week' | 'month' | 'year' | 'custom';

export interface AppState {
  intervalId?: number | null;
  today: moment.Moment;
  selectedDate: moment.Moment;
  startDate?: moment.Moment;
  endDate?: moment.Moment;
  formattedSelectedDate: string;
  pvDataCache: Map<string, MinutePVData[]>;
  totals: Map<string, Totals>;
  energyUsageLineGraphScale: Scale;
  energyUsageLineGraphView: View;
  calendarScale: CalendarScale;
}

export const initialState = (): AppState => {
  const today: moment.Moment = moment().startOf('day');
  return {
    today: today,
    selectedDate: today,
    formattedSelectedDate: dateUtils.formatDate(today),
    pvDataCache: new Map(),
    totals: new Map(),
    energyUsageLineGraphScale: 'hour',
    energyUsageLineGraphView: 'line',
    calendarScale: 'day',
  };
};