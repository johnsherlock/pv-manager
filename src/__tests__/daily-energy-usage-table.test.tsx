import { render } from '@testing-library/react';
import DailyEnergyUsageTable from '../daily-energy-usage-table';
import { EnergyCalculator } from '../lib/energy-calculator';

jest.mock('../lib/energy-calculator', () => {
  return {
    EnergyCalculator: jest.fn().mockImplementation(() => {
      return {
        recalculateTotals: jest.fn().mockReturnValue(10),
        calculateHourlyGrossCostIncStdChgAndDiscount: jest.fn().mockReturnValue(20),
        calculateDiscountedHourlyGrossCost: jest.fn().mockReturnValue(30),
        calculateExportValue: jest.fn().mockReturnValue(40),
        calculateDiscountedGrossCostExcludingStdChg: jest.fn().mockReturnValue(50),
      };
    }),
  };
});

jest.mock('../lib/num-utils', () => ({
  convertJoulesToKwh: jest.fn(),
  formatToEuro: jest.fn(),
}));

describe('DailyEnergyUsageTable', () => {
  it('renders correctly', () => {
    const data = [{ yr: 2023, mon: 2, dom: 19, hr: 1, imp: 1000, dow: 5, gep: 2000, exp: 500 }];

    const totals = {
      impTotal: 1000,
      grossCostTotal: 50,
      saturdayNetSavingTotal: 10,
      genTotal: 2000,
      grossSavingTotal: 20,
      expTotal: 500,
      exportValueTotal: 25,
    };

    const energyCalculator = new EnergyCalculator({
      dayRate: 0.4673,
      peakRate: 0.5709,
      nightRate: 0.3434,
      exportRate: 0.1850,
      discountPercentage: 0.15,
      standingCharge: 0.7066,
    });

    const component = render(
      <DailyEnergyUsageTable
        data={data}
        totals={totals}
        energyCalculator={energyCalculator}
      />);
    expect(component).toMatchSnapshot();
  });
});
