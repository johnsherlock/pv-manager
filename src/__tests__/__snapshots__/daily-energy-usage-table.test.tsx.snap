// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyEnergyUsageTable renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="table d-md-block d-none"
      >
        <div
          class="table-header"
        >
          <div
            class="table-row"
          >
            <div
              class="table-cell"
            >
              <div>
                <a
                  href="#"
                >
                  Expand
                </a>
              </div>
              <div
                class="hide"
              >
                Hour
              </div>
            </div>
            <div
              class="table-cell"
            >
              Imported kWh
            </div>
            <div
              class="table-cell"
            >
              Generated kWh
            </div>
            <div
              class="table-cell"
            >
              Consumed kWh
            </div>
            <div
              class="table-cell"
            >
              Exported kWh
            </div>
            <div
              class="table-cell"
            >
              Green Energy Coverage
            </div>
            <div
              class="table-cell"
            >
              Import Cost
            </div>
            <div
              class="table-cell"
            >
              Green Energy Saving
            </div>
            <div
              class="table-cell"
            >
              Export Value
            </div>
          </div>
        </div>
        <div
          class="table-container hide"
        >
          <div
            class="table-body"
          >
            <div
              class="table-row table-primary"
            >
              <div
                class="table-cell time"
              >
                01:30
              </div>
              <div
                class="table-cell imp"
              >
                1000.00
              </div>
              <div
                class="table-cell gep"
              >
                2000.00
              </div>
              <div
                class="table-cell comp"
              >
                1000.00
              </div>
              <div
                class="table-cell exp"
              >
                500.00
              </div>
              <div
                class="table-cell gepc"
              >
                50
                %
              </div>
              <div
                class="table-cell impCost"
              />
              <div
                class="table-cell saving"
              />
              <div
                class="table-cell expValue"
              >
                4
              </div>
            </div>
          </div>
        </div>
        <div
          class="table-footer"
        >
          <div
            class="table-row"
          >
            <span
              class="table-cell"
            >
              Total
            </span>
            <span
              class="table-cell"
            >
              8
               
              kWh
            </span>
            <span
              class="table-cell"
            >
              9
               
              kWh
            </span>
            <span
              class="table-cell"
            >
              10
               
              kWh
            </span>
            <span
              class="table-cell"
            >
              11
               
              kWh
            </span>
            <span
              class="table-cell"
            >
              12
              %
            </span>
            <span
              class="table-cell"
            >
               
            </span>
            <span
              class="table-cell"
            />
            <span
              class="table-cell"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="table d-md-block d-none"
    >
      <div
        class="table-header"
      >
        <div
          class="table-row"
        >
          <div
            class="table-cell"
          >
            <div>
              <a
                href="#"
              >
                Expand
              </a>
            </div>
            <div
              class="hide"
            >
              Hour
            </div>
          </div>
          <div
            class="table-cell"
          >
            Imported kWh
          </div>
          <div
            class="table-cell"
          >
            Generated kWh
          </div>
          <div
            class="table-cell"
          >
            Consumed kWh
          </div>
          <div
            class="table-cell"
          >
            Exported kWh
          </div>
          <div
            class="table-cell"
          >
            Green Energy Coverage
          </div>
          <div
            class="table-cell"
          >
            Import Cost
          </div>
          <div
            class="table-cell"
          >
            Green Energy Saving
          </div>
          <div
            class="table-cell"
          >
            Export Value
          </div>
        </div>
      </div>
      <div
        class="table-container hide"
      >
        <div
          class="table-body"
        >
          <div
            class="table-row table-primary"
          >
            <div
              class="table-cell time"
            >
              01:30
            </div>
            <div
              class="table-cell imp"
            >
              1000.00
            </div>
            <div
              class="table-cell gep"
            >
              2000.00
            </div>
            <div
              class="table-cell comp"
            >
              1000.00
            </div>
            <div
              class="table-cell exp"
            >
              500.00
            </div>
            <div
              class="table-cell gepc"
            >
              50
              %
            </div>
            <div
              class="table-cell impCost"
            />
            <div
              class="table-cell saving"
            />
            <div
              class="table-cell expValue"
            >
              4
            </div>
          </div>
        </div>
      </div>
      <div
        class="table-footer"
      >
        <div
          class="table-row"
        >
          <span
            class="table-cell"
          >
            Total
          </span>
          <span
            class="table-cell"
          >
            8
             
            kWh
          </span>
          <span
            class="table-cell"
          >
            9
             
            kWh
          </span>
          <span
            class="table-cell"
          >
            10
             
            kWh
          </span>
          <span
            class="table-cell"
          >
            11
             
            kWh
          </span>
          <span
            class="table-cell"
          >
            12
            %
          </span>
          <span
            class="table-cell"
          >
             
          </span>
          <span
            class="table-cell"
          />
          <span
            class="table-cell"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
